File: 1995-finding-pairs-with-a-certain-sum/finding-pairs-with-a-certain-sum.cpp
class FindSumPairs {
private:
    std::vector<int> nums1;
    std::vector<int> nums2;
    std::map<int, int> frequency;

public:
    FindSumPairs(vector<int>& nums1, vector<int>& nums2) {
        this->nums1 = nums1;
        this->nums2 = nums2;

        for (const auto& num : nums2) {
            ++frequency[num];
        }
    }
    
    void add(int index, int val) {
        int length = nums2.size();

        if(index >= 0 && index <= length-1) {
            --frequency[nums2[index]];
            nums2[index] += val;
            ++frequency[nums2[index]];
        }
    }
    
    int count(int total) {

        int countOfPairs = 0;

        for(int i = 0; i < nums1.size(); ++i) {
            int complement = total - nums1[i];

            if (frequency.find(complement) != frequency.end()) {
                countOfPairs += frequency[complement];
            }
        }

        return countOfPairs;
    }
};

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs* obj = new FindSumPairs(nums1, nums2);
 * obj->add(index,val);
 * int param_2 = obj->count(tot);
 */

File: 1995-finding-pairs-with-a-certain-sum/finding-pairs-with-a-certain-sum.java
class FindSumPairs {

    private int[] nums1;
    private int[] nums2;
    private HashMap<Integer, Integer> frequency = new HashMap<Integer, Integer>();

    // TC : O(n+m)
    public FindSumPairs(int[] nums1, int[] nums2) {
        this.nums1 = nums1; // O(n)
        this.nums2 = nums2; // O(m)

        // Update the frequency of nums2 as it is larger than nums1
        for(int nums: nums2) { // O(m)
            frequency.put(nums, frequency.getOrDefault(nums,0)+1);
        }
    }
    
    // TC : O(1)
    public void add(int index, int val) {
        // Decrease the frequency
        frequency.put(nums2[index], frequency.get(nums2[index])-1);

        // update the value from nums2
        nums2[index] += val;

        // Increse the frequency
        frequency.put(nums2[index], frequency.getOrDefault(nums2[index], 0)+1);
    }
    
    // TC : O(n)
    public int count(int tot) {
        int count = 0;

        for(int num: nums1) {
            count += frequency.getOrDefault(tot - num, 0);
        }

        return count;
    }
}

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs obj = new FindSumPairs(nums1, nums2);
 * obj.add(index,val);
 * int param_2 = obj.count(tot);
 */